packages/router/src/router.tsx

```diff
@@ -125,6 +125,7 @@ export function useHistory() {
 				! options.transition
 			) {
 				performPush();
+				return;
 			}
 
 			await new Promise< void >( ( resolve ) => {
```

Comment:
I think we can probably remove the `?? ''` fallback here

Reply:
I thought the early return will tell typescript that the classname is not empty there anymore.

Reply:
It would have been the case if the calls were synchronous, but this is called in an async function.

Reply:
Currently TypeScript cannot narrow down the type from `string | undefined` to `undefined` because the `options.transition` usage is inside a nested function (the promise constructor). In such a case it's harder to guarantee in what order code is executed.

But you could solve this by extracting most of the code out of the `new Promise` callback. The only part that needs to be async is the final `transition.finished.finally` call. Everything before it is sync.